### DISCUSSION TITLE
Learning / Reviewing the Code

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3939] [Development]
All changes to the code should go though code review.
That means opening a [Pull Request](https://github.com/neomutt/neomutt/pulls) (PR).
This gives other developers a chance to check that the code works ask questions about the implementation and spot any silly mistakes.

**This is an invitation to everyone** to ask questions in the PRs and learn about the code.

Creating a Pull Request (PR) automatically triggers lots of GitHub Actions.
They perform a suites of tests, including:

- Build NeoMutt with a variety of `configure` options
- Check the Unit Tests succeed
- Check for memory leaks ([ASAN](https://neomutt.org/dev/analysis/asan))
- Fuzz testing on the address parser
- Static "[Code Quality](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/about-code-scanning-with-codeql)" checks

For more info, see:

- [Docs for GitHub Actions](https://docs.github.com/en/actions)
- [NeoMutt's Actions (yaml config)](https://github.com/neomutt/neomutt/tree/main/.github/workflows)
- [Results of recent Actions](https://github.com/neomutt/neomutt/actions)

---

@navi-desu @roccoblues @lucilanga @g0mb4 @amudtogal
As active members, would you like to join the [Reviewers' Team](https://github.com/orgs/neomutt/teams/reviewers)?
This would mean that you are notified of every new PR.

**Comment by lucilanga** [comment]
Aye, cap'n!
I would like to join.
But then I have to pop the question, are we really qualified to be reviewers ?
Obviously I shall endeavour to become a qualified reviewer, probably everyone in this group shall do so.
But I'm not fully versed in IMAP or maildir, right now. How's that going to be handled ?
Naturally I trust everything gahr has put on a review, you .. I don't even count.
But how is that going to work for us the mortals ?

**Comment by flatcap** [comment]
> are we really qualified to be reviewers ?

Who is? :wink:
My own judgement has proven to be faulty! 

> But I'm not fully versed in IMAP or maildir, right now.
> But how is that going to work for us the mortals ?

A couple of ways.

First, it'll ensure that I actually make PRs.
In the past, sometimes I've skipped this step and had to revert stuff.

Second, more reviewers is more people asking questions.
I'll make more of an effort to explain my PRs, so you know what I'm doing and why (even if you don't yet understand the how).

> I would like to join.

done

> Obviously I shall endeavour to become a qualified reviewer

Thanks, it's really appreciated.

**Comment by auouymous** [comment]
> are we really qualified to be reviewers ?

No one is perfect, more reviewers can find more issues. You could fix something no one else would have seen by partially reviewing the changes you do understand. And testing the changes with a config or environment different from the author's can also find problems.

**Comment by roccoblues** [comment]
Thanks for the invite! Please add me too.

**Reply by flatcap** [reply]
done.  thanks!

**Comment by ossilator** [comment]
i recommend https://gerrithub.io/ if you want to go full in on code reviews.

**Comment by flatcap** [comment]
Thanks.  I'm not sure I'm ready for _that_ yet :-)

**Comment by g0mb4** [comment]
Dear @flatcap !

Thank you for the invite but I'll wait a bit to get more comfortable with the codebase and with the inner workings of the code :)

**Reply by flatcap** [reply]
sure thing

**Comment by Amudtogal** [comment]
Hey,
I feel quite honored about joining the reviewer team, and would gladly do it!
My only catch at the moment is the lingering PhD, so I don't know yet how much time I can allocate...
If that's no concern, "aye" as well from my side :smile: 

I will be diving into the completion of the autocompletion (pun intended), and will get back to you with a PR soonish!
Cheers

**Reply by flatcap** [reply]
done, thanks!
